Source/common/Keychain.mm

```diff
@@ -0,0 +1,217 @@
+}
+
+}  // namespace keychain_utils
+
+std::unique_ptr<KeychainManager> KeychainManager::Create(NSString *service,
+                                                         SecPreferencesDomain domain) {
+  if (!keychain_utils::IsValidServiceName(service)) {
+    return nullptr;
+  }
+
+  SecKeychainRef keychain = NULL;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  OSStatus status = SecKeychainCopyDomainDefault(domain, &keychain);
+#pragma clang diagnostic pop
+  if (status != errSecSuccess || keychain == nullptr) {
+    LOGE(@"Failed to get desired keychain. Domain: %d, status: %d", domain, status);
+    return nullptr;
+  }
+
+  return std::make_unique<KeychainManager>(service, keychain);
+}
+
+KeychainManager::KeychainManager(NSString *service, SecKeychainRef keychain)
+    : service_(service), keychain_(keychain) {
+  assert(keychain_ != nullptr);
+}
+
+KeychainManager::~KeychainManager() {
+  if (keychain_ != nullptr) {
+    CFRelease(keychain_);
+  }
+}
+
+KeychainManager::KeychainManager(KeychainManager &&other)
+    : service_(std::move(other.service_)), keychain_(other.keychain_) {
+  other.service_ = nil;
+  other.keychain_ = nullptr;
+}
+
+KeychainManager &KeychainManager::operator=(KeychainManager &&other) {
+  if (this != &other) {
+    if (keychain_ != nullptr) {
+      CFRelease(keychain_);
+    }
+
+    service_ = std::move(other.service_);
+    keychain_ = other.keychain_;
+
+    other.service_ = nil;
+    other.keychain_ = nullptr;
+  }
+  return *this;
+}
+
+std::unique_ptr<KeychainItem> KeychainManager::CreateItem(NSString *account,
+                                                          NSString *description) {
+  if (!keychain_utils::IsValidAccountName(account)) {
+    LOGE(@"Invalid account name for keychain item: %@", account);
+    return nullptr;
+  }
+
+  if (!keychain_utils::IsValidDescription(description)) {
+    LOGE(@"Invalid description for keychain item: %@", description);
+    return nullptr;
+  }
+
+  return std::make_unique<KeychainItem>(service_, account, description, keychain_);
+}
+
+KeychainItem::KeychainItem(NSString *service, NSString *account, NSString *description,
+                           SecKeychainRef keychain)
+    : service_(service), account_(account), description_(description), keychain_(keychain) {
+  assert(keychain_ != nullptr);
+  // Retain the keychain reference for ourselves
+  CFRetain(keychain_);
+}
+
+KeychainItem::~KeychainItem() {
+  if (keychain_ != nullptr) {
+    CFRelease(keychain_);
+  }
+}
+
+absl::Status KeychainItem::Store(NSData *data) {
+  if (data.length == 0) {
+    return absl::ErrnoToStatus(EINVAL, "No data to store");
+  }
+
+  if (auto status = Delete(); !status.ok()) {
+    LOGE(@"Failed to remove previous value. %s", status.message().data());
+    return status;
+  }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
```

Comment:
Which of these are deprecated? Do we really need the deprecated ones?

Reply:
`kSecAttrAccessibleAlwaysThisDeviceOnly` is "deprecated", but mainly seems to be in the sense to communicate that it shouldn't be used outside of launch daemons as far as I can tell. We don't have much of an option here.